### ISSUE TITLE
Add a keyboard shortcut to toggle the master JavaScript switch

### ISSUE THREAD

**Issue by ghost** [uBlockOrigin/uBlock-issues#692]
ABO need an more usable temporary javascript toggle, I suggest, add an keyboard shortcut to toogle "Javascript on this site" and reload page. Why not add this little and easy and very useful thing?

**Comment by uBlock-user** [comment]
Duplicate of https://github.com/gorhill/uBlock/issues/753

**Comment by gorhill** [comment]
I am going to re-open this because I do not consider the original to be a duplicate, as there was no master JavaScript switch back then.

**Comment by mrochaaa** [comment]
@gorhill, thanks, please add some shortcuts to us, like to master js, local images etc, and think about add also an advanced option to auto reload page after an keyboard shortcut toggle to reduce the number of clicks. I suggest add this advanced option to auto reload when toggle using keyboard or toggle direct in panel, must be just one.

AutoReloadPage = true

**Comment by gorhill** [comment]
Please do not flood the repo with requests, including adding more request to existing issues.

I merely re-opened the issue here because I want to give some more thoughts about whether an additional shortcut for the JavaScript master switch is really desirable -- I have not come to a decision yet.

Auto-reload when clicking has been ruled out already in other issues in the past, do not keep asking for it. Auto-reload already exists for the "Relax blocking mode" shortcut because in that case it makes sense.

**Comment by mrochaaa** [comment]
All right, absolutely this is my last suggestion for ublock, I'm sure which many of my suggestions can be very useful for many, I have many friends with the same needs, also have other issues opened in the past, but the developers don't care, I think ublock is very independent and mature, no need any new thing to be better, is almost perfect, and who care about 10 clicks for each little change the panel? Absolutely anyone! Sorry for the food and all my ridiculous ideas, I'm just an noob.

Thanks so much

**Comment by gorhill** [comment]
> who care about 10 clicks for each little change in the panel?

If you want to convince me you need better arguments than sarcasm.

I am a user of advanced mode since the beginning (years), and auto-reload has never been something I wanted or ever felt was needed. I always want the reload to occur when I tell the extension to do so, never when the extension decide to do so, not when the extension seemingly arbitrarily decides to do so, and especially not for every single click.

**Comment by mrochaaa** [comment]
@gorhill you is an advanced user, but you is not the unique type of advanced user, saying with all humble, my first "white" PC had DOS, my FF now have just 60 essential addons, counting. For instance I use javascript disabled for all sites, and I enable selectively, today I need  4 clicks to do it, 1 for open the panel, one on js toggle, one to reload, one to close panel, so, I could do it with just 1 click to allow js temporary, I think ublock can even add additional extensions, to add exclusive toggle buttons to toolbar, to js, images , "helpers", to toggle with 1 click, like in the past Adblock had Element Hide Helper, today I have ublock with have image and js block, but I need 2 other plugins to block images and javascripts, because "to me" to "my type" of advanced user 4 clicks each change is much, I'm not only one which think so, why have a lot of users which used "All in One Sidebar" to use the "Auto Hide"? To avoid clicks! My sidebar still with auto hide today. Many want Auto Reload, but not by default, but an optional and advanced setting, why not? Use it who want, the same thing is Panel Auto Close and the "helpers" install the helper who want and need, 

I think which ublock is good but need much changes and much evolution in the "usability" have a lot of power, but today is hard to use his power, forcing users to use other plugins just to get "usability" to get "one click". And have a lot of "opened issues" asking this "easy usability" in many ways and different words.

Is my opinion.

**Comment by gorhill** [comment]
> I use javascript disabled for all sites, and I enable selectively, today I need 4 clicks to do it

Why not use the "Relax blocking mode" shortcut? It does enable JavaScript and auto-reload the page.

**Comment by mrochaaa** [comment]
surely is the best option today, but for instance with "helpers addons" the user could see the status of js in the icon, and even re-disable js (real toggle), and also use click instead of keyboard. In an another "issue" I suggested an message receiver (as other users suggested), so the user can toggle using an Foxy Gesture script. I suggest also an "Relax" to images, even better an selective image control, like the legacy addon "image like opera". So user can load just the wanted image with CTRL + click, and add settings to don't block small images, add an white list, to allow captchas etc. So, ublock have an powerful image blocker, but with an weak usability and interface, but with few new settings and controls you can have an complete image control using just one blocking engine. Just to make clear, image block is essential in some countries with limited bandwidth.

**Comment by gorhill** [comment]
> re-disable js

This is the issue here.

> user can toggle using an Foxy Gesture script

I declined this already, as explained in the original issue.

> user can load just the wanted image with CTRL + click

You can use "Block large media elements" for such functionality.

**Comment by mrochaaa** [comment]
For instance if I use an Global Rule to block images for all sites, and I want unblock images in the current tab, need 4 clicks. Need at least and new shortcut, an relax to images.

**Comment by B83C** [comment]
> 

How do you go about setting it up? Doesn't seem to work here... 
Triggering relax blocking mode allows only going from hard to easy mode but not the other way.
Do you have any solutions?

**Comment by B83C** [comment]
gg

> > I use javascript disabled for all sites, and I enable selectively, today I need 4 clicks to do it
> 
> Why not use the "Relax blocking mode" shortcut? It does enable JavaScript and auto-reload the page.
How do you go about setting it up? Doesn't seem to work here...
Triggering relax blocking mode allows only going from hard to easy mode but not the other way.
Do you have any solutions?

**Comment by gwarser** [comment]
@B83C https://github.com/uBlockOrigin/uBlock-issues/wiki/Advanced-settings#blockingprofiles

**Comment by B83C** [comment]
So basically, there's no way to toggle it in a cyclic manner?

**Comment by julesallen** [comment]
Been watching this one for a while and it doesn't look like we're getting a keystroke to toggle Javascript. Really appreciate uBO @gorhill, it's your extension, your rules, and respect you get final say. Like others on here I'd love a toggle to be built in, I drive around with scripting turned off by default.

My embarrassing hack to get this to work is a filthy, horrible macro running on [Keyboard Maestro](https://www.keyboardmaestro.com/main/) on macOS. It requires the uBo icon to be in the same fixed place.

<img width="191" alt="Screenshot 2023-07-15 at 12 44 29" src="https://github.com/uBlockOrigin/uBlock-issues/assets/871038/8617614e-93fb-4994-a586-d812e148ada1">

In pseudo code, it goes like this:

```
if the browser is at the front:
    store the mouse position
    wait .2 seconds # if the interface needs to be loaded into memory
    move (from the position of the top right of the window) and click the uBo icon
    wait .15 seconds
    move (relatively) and click the javascript toggle
    wait .1 seconds
    move (relatively) and click reload
    wait .1 seconds
    send the escape key to close the uBo window
    restore the mouse position
else:
    quit
```

Not my proudest programming moment but this will toggle .js on and off with a keystroke. Shave down the timing if you've got a better machine than mine and about half a second is too long for you.

Or use whatever macro editor you prefer.

[Toggle Javascript uBlock.kmmacros.zip](https://github.com/uBlockOrigin/uBlock-issues/files/12062427/Toggle.Javascript.uBlock.kmmacros.zip) <-- XML macro file

Not everybody has Keyboard Maestro. Maybe have [Automator](https://support.apple.com/guide/automator/welcome/mac) record your actions, save it as an app, and call with a keystroke.

<img width="676" alt="automator-record" src="https://github.com/uBlockOrigin/uBlock-issues/assets/871038/e4bf997a-7d5b-4fd9-bfaf-323a5c7b3d9e">

Or use whatever macro editor you prefer — all ears if you've got a better solution.

**Comment by landscapejohn** [comment]
The biggest use case I have for this right now is that I'm on a page that uses mouse hovers to popup modal menus. I want to block some cosmetic elements in one of the modals but it disappears as soon as focus is shifted away to the UBO element picking window.

I thought I could just inspect element and pick it up there but I'm running into the same problem where the element goes away as soon as I move the mouse to expand the div and see what tags apply to the element I want to block.

It would be really cool if I could hover over the element, get the modal popup, then hit a key to disable javascript thus leaving the modal on the screen and being able to inspect/block what elements it contains.

**Comment by julesallen** [comment]
> It would be really cool if I could hover over the element, get the modal popup, then hit a key to disable javascript thus leaving the modal on the screen and being able to inspect/block what elements it contains.

That's a use case I hadn't though of. I wonder if calling the extension might lose focus on the rendered page you're viewing?

This sometimes works:

1. Make a mental note of where the cursor is.
2. Win/⌘ tab away to another app.
3. Move the cursor to the top of the screen, Win/⌘ tab back to browser.
4. Turn off Javascript in uBO, Win/⌘ tab away.
5. Move the cursor back to where it was.
6. Win/⌘ tab back to browser.

I haven't done step 4 with uBO, for me it's switch to an editor, make some changes, move back without changing focus. If you use a macro program it makes this 100% precise without having to guess where the cursor goes back to.

**Comment by gorhill** [comment]
> That's a use case I hadn't though of

That's not how it works anyway, the no-scripting switch does not stop JS execution, it prevents it from starting. Once it's started, uBO can't pause it.

**Comment by gwarser** [comment]
Does not work in Firefox

**Comment by gorhill** [comment]
Right, forgot to modify other `manifest.json`s
